File: src/plugin/CloudPlugin.php
<?php

namespace pocketcloud\cloud\plugin;

use pocketcloud\cloud\console\log\logger\PrefixedLogger;
use pocketcloud\cloud\Server;
use pocketcloud\cloud\scheduler\TaskScheduler;
use pocketcloud\cloud\util\FileUtils;
use pocketcloud\cloud\util\PathUtils;

abstract class CloudPlugin {

    private bool $enabled = false;
    private TaskScheduler $scheduler;
    private PrefixedLogger $prefixedLogger;

    public function __construct(
        private readonly CloudPluginDescription $description,
        private readonly string $dataFolder,
        private readonly string $pluginDirPath
    ) {
        $this->scheduler = new TaskScheduler($this);
        $this->prefixedLogger = new PrefixedLogger($Server::getInstance()->logger, "[" . $this->description->getName() . "]");
    }

    public function saveResource(string $relativePath, bool $replace = false): bool {
        $absoluteSourcePath = PathUtils::join($this->pluginDirPath, "resources", $relativePath);
        $absoluteDestinationPath = PathUtils::join($this->dataFolder, $relativePath);
        if (!file_exists($absoluteSourcePath)) return false;
        if (file_exists($absoluteDestinationPath) && !$replace) return false;
        if (!file_exists(dirname($absoluteDestinationPath))) FileUtils::createDir(dirname($absoluteDestinationPath));
        return copy($absoluteSourcePath, $absoluteDestinationPath);
    }

    public function saveDefaultConfig(): bool {
        return $this->saveResource("config.yml");
    }

    public function onLoad(): void {}

    public function onEnable(): void {}

    public function onDisable(): void {}

    public function setEnabled(bool $enabled): void {
        $this->enabled = $enabled;
    }

    public function isEnabled(): bool {
        return $this->enabled;
    }

    public function isDisabled(): bool {
        return !$this->enabled;
    }

    public function getScheduler(): TaskScheduler {
        return $this->scheduler;
    }

    public function getLogger(): PrefixedLogger {
        return $this->prefixedLogger;
    }

    public function getDescription(): CloudPluginDescription {
        return $this->description;
    }

    public function getDataFolder(): string {
        return $this->dataFolder;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/plugin/CloudPlugin.java
package de.pocketcloud.cloud.plugin;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.console.log.def.PrefixedLogger;
import de.pocketcloud.cloud.scheduler.TaskScheduler;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.Nullable;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

@Getter
public abstract class CloudPlugin {

    private boolean initialized = false;

    @Setter
    private CloudPluginState state = CloudPluginState.DISABLED;
    private CloudPluginClassLoader loader;
    private TaskScheduler scheduler;
    private PrefixedLogger logger;
    private CloudPluginDescription description;
    private Path dataFolder;
    private Path pluginFilePath;

    public CloudPlugin() {}

    public final void init(CloudPluginClassLoader loader, CloudPluginDescription description, Path dataFolder, Path pluginFilePath) {
        if (initialized) return;
        initialized = true;
        this.loader = loader;
        this.scheduler = new TaskScheduler(this);
        this.description = description;
        this.dataFolder = dataFolder;
        this.pluginFilePath = pluginFilePath;
        this.logger = CloudLogger.prefixed("[" + description.name() + "]");
    }

    public boolean saveResource(String relativePath, boolean overwrite) throws IOException {
        if (relativePath == null || relativePath.isBlank()) throw new IllegalArgumentException("relativePath cannot be null or blank");
        if (relativePath.contains("..")) throw new IllegalArgumentException("Path traversal is not allowed");
        Path destination = dataFolder.resolve(relativePath);
        if (Files.exists(destination) && !overwrite) return false;
        Files.createDirectories(destination.getParent());
        try (InputStream entryInputStream = getResource(relativePath)) {
            if (entryInputStream == null) return false;
            if (overwrite) {
                Files.copy(entryInputStream, destination, StandardCopyOption.REPLACE_EXISTING);
            } else {
                Files.copy(entryInputStream, destination);
            }
        }
        return true;
    }

    public void onLoad() {}

    public void onEnable() {}

    public void onDisable() {}

    public boolean isEnabled() {
        return state == CloudPluginState.ENABLED;
    }

    public boolean isDisabled() {
        return state == CloudPluginState.DISABLED;
    }

    @Nullable
    public InputStream getResource(String path) throws IOException {
        try (JarFile jarFile = new JarFile(pluginFilePath.toFile())) {
            JarEntry entry = jarFile.getJarEntry(path);
            if (entry == null) return null;
            try (InputStream stream = jarFile.getInputStream(entry)) {
                return new ByteArrayInputStream(stream.readAllBytes());
            }
        }
    }

    final public PocketCloud getCloud() {
        return PocketCloud.instance();
    }
}